### DISCUSSION TITLE
Composite key support

### DISCUSSION THREAD

**Discussion by noedelcroix** [directus/directus#12137] [Feature Requests]
Hello,
A really missing functionality support is the composite management. Directus doesn't permit to create a unique composite key from GUI and even we create the composite key from SQL command (with ```ALTER TABLE <table name> ADD UNIQUE <key name>(<indexes>);```), no errors are thrown...

**Comment by NilsBaumgartner1994** [comment]
I would like to see unique composite keys.

For example: A user can set allergic markings, which should not be dublicate in the junction table

**Reply by u12206050** [reply]
The easiest and simplest way to do this is adding a constraint on the two columns that should be unique.
The reason a single will always be preferred over two is both for navigation (CRUD) purposes as well as keeping things simple.

If you want here is a simple migration for creating a composite index on two columns (OPTIONALLY MORE).
```
// extensions/migrations/20220625-somethings-composite-index.js

module.exports = {
    async up(knex) {
        await knex.schema.table('somethings', (table) => {
            table.unique(['thing', 'other']);
        });
    },

    async down(knex) {
        await knex.schema.table('somethings', (table) => {
            table.dropUnique(['thing', 'other']);
        });
    },
};
```

**Comment by NilsBaumgartner1994** [comment]
How would I create the constraint in the GUI?


Beste Grüße,
Nils Baumgartner

**Reply by rijkvanzanten** [reply]
There's currently no support for composite unique/foreign/primary keys 👍🏻

**Reply by ozum** [reply]
@rijkvanzanten, as I advocated at #17185, this should be considered a bug, because composite keys are part of SQL standard and this is a blocking issue for databases with composite primary keys.

**Comment by cube-dan** [comment]
I needed a composite unique key across two M2M fields. To get around this missing feature I ended up writing a [filter() hook implementation](https://docs.directus.io/extensions/hooks.html#filter). Here's some example code: 

`src/index.js` within the [generated extension](https://docs.directus.io/extensions/creating-extensions.html#creating-extensions)

```javascript
// a custom re-implementation of the Directus RecordNotUniqueException; see below.
import { RecordNotUniqueException } from './RecordNotUniqueException';

export default ({ filter }, { exceptions, services }) => {

  const { ItemsService } = services;

  filter('assignment.items.create', async (payload, collection, ctx) => {
    
    const { accountability, schema } = ctx;
    const Assignment = new ItemsService('assignment', { schema: schema, accountability: accountability });
    
    // get the two M2M field values from the incoming payload
    const {item_id, profile_id} = payload;

    // do we already have an existing [profile]>-<[item] combination?
    const assignment = await Assignment.readByQuery({
      filter: {
        item_id: {
          _eq: item_id,
        },
        profile_id: {
          _eq: profile_id,
        },
      },
    });

    // stop the creation of the record if an existing [profile_id]>-<[item_id] record already exists
    if(typeof assignment[0] !== 'undefined'){
      throw new RecordNotUniqueException('[item]>-<[profile]]');
    }

  });

};
```

`src/RecordNotUniqueException.js` — 'cause RecordNotUniqueException isn’t included in the [exceptions that directus passes down](https://docs.directus.io/extensions/hooks.html#filter) through { exceptions }

```javascript
// @SOURCE: node_modules/directus/dist/exceptions/database/record-not-unique.js
import { BaseException } from '@directus/shared/exceptions';
class RecordNotUniqueException extends BaseException {
    constructor(field, extensions) {
        if (field) {
            super(`Field "${field}" has to be unique.`, 400, 'RECORD_NOT_UNIQUE', extensions);
        }
        else {
            super(`Field has to be unique.`, 400, 'RECORD_NOT_UNIQUE', extensions);
        }
    }
}
export { RecordNotUniqueException };
```

**Comment by Natetronn** [comment]
We created a gist explaining how we did something similar using Flows:

https://gist.github.com/Natetronn/9bafd1722a2b67b9bf4528e45992baa2

Note: it may be a bit rough around the edges and probably not the best method of achieving this, but perhaps it can be of use to others, just the same.

**Reply by geekyayush** [reply]
You are a life saver. Thank you.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by realjjaveweb** [reply]
Hi! May I ask what is the current state?
My 2 cents: compound primary keys should be at least supported for basic M:N "pairing" tables; adding extra row ID would be only acceptable if there would be more data to the relationship (although it may be considered breaking some of the normal forms), however if there would be more remove/add it would mean unnecessary incrementing.

Some ORMs also create such tables (only 2 columns M_id & N_id as compound primary key) by default for M:N relationships.

**Reply by rijkvanzanten** [reply]
It's being worked on as part of the bigger data pipeline restructuring 👍🏻

**Reply by realjjaveweb** [reply]
Sounds great! :blush:  is there an approximate release date ? :slightly_smiling_face:

**Reply by rijkvanzanten** [reply]
Hard to estimate seeing it's such a big project! Keep an eye on https://github.com/directus/directus/issues/19762 though for all updates on that front 👍🏻

**Reply by divStar** [reply]
#19762 sounds like a monumental task (judging by the experience I have myself as a software engineer). I suppose it needs to be completed before we see any progress on composite IDs / keys?

I just have a collection where I have a `position`, a `skill` and a `skill-level` and I'd like to disallow adding new elements if the former two are already present in the collection (regardless of the value of the third field) and I am a bit puzzled, that this feature is missing while many other highly complex features are actually present AND working flawlessly.

**Reply by rijkvanzanten** [reply]
> https://github.com/directus/directus/issues/19762 sounds like a monumental task (judging by the experience I have myself as a software engineer). I suppose it needs to be completed before we see any progress on composite IDs / keys?

It is, _but_ refactoring the current data pipeline to support composite keys is similarly a herculean task, hence my preference of including it in the new abstraction we're building rather then trying to spend all that effort on it in the current setup and then completely replacing that right after 🙂 

> and I am a bit puzzled, that this feature is missing while many other highly complex features are actually present AND working flawlessly.

Fair enough! That mostly stems from Directus not having those other highly complex features in the past either. The original API design really stems from that `/:table` / `/:table/:id` paradigm, where you only really have a single primary key just by nature of REST API design. In 2024 that's completely different!

**Reply by realjjaveweb** [reply]
@rijkvanzanten I see you're a person on the right place, **if** we get to see composite keys support **some** day in relatively near future (like a year or so max), I totally agree that it makes sense to only do it in the new abstraction and don't waste time on the old one :slightly_smiling_face:

**Reply by cmann90** [reply]
> Hard to estimate seeing it's such a big project! Keep an eye on https://github.com/directus/directus/issues/19762 though for all updates on that front 👍🏻

Good to see that an implementation is planned. The shared link however is not valid anymore. I also didn't find a replacement for it.

![image](https://github.com/user-attachments/assets/27920655-b584-440d-a54a-c86cb93261f0)

**Reply by kgamora** [reply]
@rijkvanzanten , is there a way to track the feature request status?

**Comment by u12206050** [comment]
As a simple workaround we have created key constraints on two columns being unique via a migration. This does throw and error in the admin dashboard if we attempt to create a record with two keys that are already paired.
```
// migrations/20221010-links-composite.ts

import { Knex } from 'knex';

export async function up(knex: Knex) {
  await knex.schema.table('links', table => {
    table.unique(['person', 'other']);
  });
}

export async function down(knex: Knex) {
  await knex.schema.table('links', table => {
    table.dropUnique(['person', 'other']);
  });
}

```

**Comment by FreeSynergy** [comment]
Would it help, to bring more voters for this idea to get a higher priority?

**Reply by rijkvanzanten** [reply]
That always helps! While I can't just go top-to-bottom all the time — gotta choose my battles wisely — I do use the upvote count and 'engagement' as a first pass of priority as communicated by the Directus community overall 🙂

**Comment by edubacco** [comment]
I think directus definitely needs this feature, as without it it's quite difficult to use directus on top of existing relational databases

**Reply by kgamora** [reply]
It's actually hard to use them as-is because there's no clear way to save the configuration.

I tried using `directus-sync` but stuck on issue that it treats existing tables as collections from the start and thinks it needs to create them, hence running into an error. I opened the issue: https://github.com/tractr/directus-sync/issues/202

**Comment by jclaveau** [comment]
I opened a feature request to be able to cancel an item creation from a filter hook or **to replace the creation result by an existing one**: https://github.com/directus/directus/discussions/25411

This made me able to deduplicate items in complex chained m2m relations based on composite unique PG index

I use the class `Model` below the following way:
```ts
const tableName = `some_collection_with_compiste_index`
type AggregateType = SomeCollectionWithCompositeIndex

export default defineHook(({ filter, action }, hookContext: DirectusHookContext) => {
  const model = new Model<AggregateType>(tableName, hookContext)

  filter<AggregateType>(`${tableName}.items.create`, async (
    input,
    meta,
    ctx
  ) => {
    const {
      updatedPrimaryKey,
    } = await model.updateIfDuplicated(
      input,
      ctx
    )
    if (updatedPrimaryKey) {
      return updatedPrimaryKey
    }

    return input
  })
})
```


```ts
// This implementation is for Directus 10.4.3 which I'm currently upgrading
import type {
  DirectusHookContext,
  EventContext,
} from '../types'
import { Item, ItemInput } from '@directus/sdk'
export type PrimaryKey = string | number; // Taken from items.ts

import {
  inputToQueryFilter,
  filterTreeToAnd,
} from './directus-lib'

export class Model<ModelType extends Item> {
  constructor(
    private tableName: string,
    private hookContext: DirectusHookContext
  ) { }

  getService(eventContext: EventContext) {
   const currentAggregateService = new this.hookContext.services
    .ItemsService<ModelType>(
      this.tableName,
      {
        schema: eventContext.schema as Exclude<typeof eventContext.schema, null>,
        accountability: eventContext.accountability,
      },
    )

    return currentAggregateService
  }


  async listUniqueIndexes(eventContext: EventContext) {
    // TODO
    // - memoize the indexes
    // - use knex-schema-inspector fork
    const rawSql = `
      select 
        pg_indexes.tablename,
        pg_indexes.indexname,
        pg_indexes.tablespace,
        pg_indexes.indexdef
      from pg_indexes 
      where 
        tablename = '${this.tableName}'
    `
    const indexesResponse = await eventContext.database.raw<{
      rows: {
        tablename: string
        indexname: string
        tablespace: string
        indexdef: string
      }[]
    }>(rawSql)

    return indexesResponse
    .rows
    .filter((index) => index.indexdef.match(/^CREATE UNIQUE INDEX /))
    .map((index) => {
      const matches = index.indexdef.match(/\(([^)]+)\)$/)
      if (! matches) {
        throw new Error(`Invalid indexdef while searching unique indexes: ${index.indexdef}`)
      }
      const fields = matches[1]?.split(/\s*,\s*/)
      .filter((field) => field.length)

      return fields
      // ?.console('listIndexes() index fields')
    })
    // .console(`listIndexes() ${this.tableName} indexesFields`)
  }


  async listDuplicates(
    input: ItemInput<ModelType>,
    eventContext: EventContext
  ) {
    const service = this.getService(eventContext)
    const indexesFields = await this.listUniqueIndexes(eventContext)

    if (! indexesFields.length) {
      return []
    }

    if (indexesFields.length === 1
      && indexesFields[0].length === 1
      && indexesFields[0][0] === `id` // TODO id may not be the pkey
    ) {
      // No duplicate possible
      return []
    }

    // Transform the indexes into a filter
    const filters = indexesFields
    .filter((indexFields) => {
      if (! indexFields) {
        throw new Error(`Invalid indexFields: ${JSON.stringify(indexFields)}`)
      }
      const inputKeys = Object.keys(input)
      const indexFieldsIncludesOneInputKey = indexFields
      ?.some((field) => inputKeys.includes(field))

      // TODO What if there is another filter
      // ensuring unicity like pkey check first?
      const indexFieldsInInputKey = indexFields
      ?.filter((field) => inputKeys.includes(field))

      if (indexFieldsIncludesOneInputKey
          && indexFieldsInInputKey.length !== indexFields.length
      ) {
        throw new Error(
          `Some fields required by the unique index ar missing for '${this.tableName}': ${JSON.stringify({
            indexFieldsInInputKey,
            indexFields,
          }, null, 2)}`
        )
      }
      return indexFieldsIncludesOneInputKey
    })
    .map((uniqueIndexFields) => {
      const acc: { [key: string]: unknown } = {} // TODO better typing
      const filter = uniqueIndexFields
      ?.reduce((acc, field) => {
        acc[field] = inputToQueryFilter(input[field])
        return acc
      }, acc)
      return filter
    })

    if (! filters.length) {
      return []
    }


    const query = {
      // input to fields?
      // fields: inputToQueryFieldsIds(input),
      filter: filterTreeToAnd({ // filterTreeToAnd required to fix a bug in Directus's runAST()
        _and: filters,
      }),
      limit: 2, // Needed to check for existing duplicates
    }

    const duplicates = await service.readByQuery(query)

    if (duplicates.length > 1) {
      console.error(`duplicates.length > 1 query`, JSON.stringify({
        query,
        duplicatesCount: duplicates.length,
      }, null, 2))
      // console.log('duplicates', duplicates)
      throw new Error(`duplicates.length > 1 ${this.tableName}:`)
    }

    return duplicates
  }

  getPrimaryKeyField(
    input: ItemInput<ModelType>,
    eventContext: EventContext
  ) {
    const primaryKeyField = eventContext.schema
    .collections[this.tableName]!.primary
    return primaryKeyField
  }


  getPrimaryKey(
    input: ItemInput<ModelType>,
    eventContext: EventContext
  ) {
    const primaryKeyField = this.getPrimaryKeyField(input, eventContext)
    const primaryKey = input[primaryKeyField] as PrimaryKey
    return primaryKey
  }


  async updateIfDuplicated(
    input: ItemInput<ModelType>,
    eventContext: EventContext
  ) {
    const duplicates = await this.listDuplicates(input, eventContext)
    if (duplicates.length) {
      if (! duplicates[0]) {
        throw new Error(`duplicates[0] is null`)
      }
      // TODO update or throw error if there is a diff?

      const primaryKeyField = this.getPrimaryKeyField(input, eventContext)
      const primaryKey = this.getPrimaryKey(
        duplicates[0],
        eventContext
      )

      await this.update({
        ... {
          [primaryKeyField]: primaryKey,
        },
        ...input,
      }, eventContext)

      return {
        updatedPrimaryKey: primaryKey,
        duplicates,
      }
    }

    return {
      updatedPrimaryKey: null,
      duplicates: null,
    }
  }


  async update(
    input: ItemInput<ModelType>,
    eventContext: EventContext
  ) {
    const service = this.getService(eventContext)
    const primaryKey = this.getPrimaryKey(
      input,
      eventContext
    )

    if (! primaryKey) {
      throw new Error(`primaryKey is null`)
    }

    return await service.updateOne(
      primaryKey,
      input,
    )
  }
}
```


This experiment works for Postrgres only as Knex doesn't provide a unified way of retrieving indexes but it works like a charm for a while now
